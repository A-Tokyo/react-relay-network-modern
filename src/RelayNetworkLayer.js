/* @flow */

import { Network } from 'relay-runtime';
import RelayRequest from './RelayRequest';
import fetchWithMiddleware from './fetchWithMiddleware';
import type {
  Middleware,
  FetchFunction,
  SubscribeFunction,
  RNLExecuteFunction,
} from './definition';

export default class RelayNetworkLayer {
  _middlewares: Middleware[];
  execute: RNLExecuteFunction;
  fetchFn: FetchFunction;
  subscribeFn: ?SubscribeFunction;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];
    this.subscribeFn = subscribeFn;

    this.fetchFn = (operation, variables, cacheConfig, uploadables) => {
      const req = new RelayRequest(operation, variables, cacheConfig, uploadables);
      return fetchWithMiddleware(req, this._middlewares);
    };

    this.execute = Network.create(this.fetchFn, this.subscribeFn);
  }
}
