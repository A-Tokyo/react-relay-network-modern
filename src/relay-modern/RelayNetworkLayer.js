/* @flow */

import { Network } from 'relay-runtime';
import type { Middleware } from '../definition';

export type Variables = { [name: string]: $FlowFixMe };
export type ConcreteBatch = any;
export type CacheConfig = {
  force?: ?boolean,
  poll?: ?number,
  rerunParamExperimental?: ?any,
};
export type Disposable = { dispose(): void };
export type Uploadable = File | Blob;
export type UploadableMap = { [key: string]: Uploadable };
export type PayloadData = { [key: string]: mixed };
export type QueryPayload = {|
  data?: ?PayloadData,
  errors?: Array<any>,
  rerunVariables?: Variables,
|};
// this is workaround should be class from relay-runtime/network/RelayObservable.js
export type RelayObservable<T> = Promise<T>;
// Note: This should accept Subscribable<T> instead of RelayObservable<T>,
// however Flow cannot yet distinguish it from T.
export type ObservableFromValue<T> = RelayObservable<T> | Promise<T> | T;
export type FetchFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  uploadables: ?UploadableMap
) => ObservableFromValue<QueryPayload>;
// See SubscribeFunction type declaration in relay-runtime/network/RelayNetworkTypes.js
export type SubscribeFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  observer: any
) => RelayObservable<QueryPayload> | Disposable;

export default class RelayModernNetworkLayer {
  _middlewares: Middleware[];
  execute: (
    operation: ConcreteBatch,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: ?UploadableMap
  ) => RelayObservable<QueryPayload>;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];

    const fetchFn: FetchFunction = (operation, variables, cacheConfig, uploadables) => {
      return Promise.resolve({});
    };

    this.execute = Network.create(fetchFn, subscribeFn);
  }
}
