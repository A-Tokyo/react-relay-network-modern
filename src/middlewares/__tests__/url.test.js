/* @flow */

import fetchMock from 'fetch-mock';
import RelayNetworkLayer from '../../RelayNetworkLayer';
import { mockReq } from '../../__mocks__/mockReq';
import urlMiddleware from '../url';

describe('middlewares/url', () => {
  beforeEach(() => {
    fetchMock.restore();
  });

  it('`url` option as string', async () => {
    fetchMock.mock({
      matcher: '/some_url',
      response: {
        status: 200,
        body: { data: 'PAYLOAD' },
      },
      method: 'POST',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: '/some_url',
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });

  it('`url` option as thunk', async () => {
    fetchMock.mock({
      matcher: '/thunk_url',
      response: {
        status: 200,
        body: { data: 'PAYLOAD2' },
      },
      method: 'POST',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: () => '/thunk_url',
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD2');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });

  it('`method` option', async () => {
    fetchMock.mock({
      matcher: '/get_url',
      response: {
        status: 200,
        body: { data: 'PAYLOAD3' },
      },
      method: 'GET',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: '/get_url',
        method: 'GET',
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD3');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });

  it('`headers` option as Object', async () => {
    fetchMock.mock({
      matcher: '/headers_url',
      response: {
        status: 200,
        body: { data: 'PAYLOAD4' },
      },
      method: 'POST',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: '/headers_url',
        headers: {
          'custom-header': '123',
        },
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD4');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });

  it('`headers` option as thunk', async () => {
    fetchMock.mock({
      matcher: '/headers_thunk',
      response: {
        status: 200,
        body: { data: 'PAYLOAD5' },
      },
      method: 'POST',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: '/headers_thunk',
        headers: () => ({
          'thunk-header': '333',
        }),
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD5');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });

  it('`credentials` option', async () => {
    fetchMock.mock({
      matcher: '/credentials_url',
      response: {
        status: 200,
        body: { data: 'PAYLOAD6' },
      },
      method: 'POST',
    });

    const rnl = new RelayNetworkLayer([
      urlMiddleware({
        url: '/credentials_url',
        credentials: 'same-origin',
      }),
    ]);
    const req = mockReq();
    const res = await req.execute(rnl);
    expect(res.data).toBe('PAYLOAD6');
    expect(fetchMock.lastOptions()).toMatchSnapshot();
  });
});
