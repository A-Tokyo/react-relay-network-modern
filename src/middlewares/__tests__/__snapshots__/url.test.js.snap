// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/url \`credentials\` option 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {},
  "method": "POST",
  "url": "/credentials_url",
}
`;

exports[`middlewares/url \`headers\` option as Object 1`] = `
Object {
  "headers": Object {
    "custom-header": "123",
  },
  "method": "POST",
  "url": "/headers_url",
}
`;

exports[`middlewares/url \`headers\` option as thunk 1`] = `
Object {
  "headers": Object {
    "thunk-header": "333",
  },
  "method": "POST",
  "url": "/headers_thunk",
}
`;

exports[`middlewares/url \`method\` option 1`] = `
Object {
  "headers": Object {},
  "method": "GET",
  "url": "/get_url",
}
`;

exports[`middlewares/url \`url\` option as string 1`] = `
Object {
  "headers": Object {},
  "method": "POST",
  "url": "/some_url",
}
`;

exports[`middlewares/url \`url\` option as thunk 1`] = `
Object {
  "headers": Object {},
  "method": "POST",
  "url": "/thunk_url",
}
`;
